tc_portal_sync.py

```diff
@@ -0,0 +1,472 @@
+#!/usr/bin/env python
+import argparse
+import copy
+import json
+import logging.config
+import socket
+import sys
+import time
+from datetime import datetime
+from datetime import timezone
+from pathlib import Path
+
+import boto3
+import botocore.exceptions
+import certifi
+import dateutil
+import dateutil.parser
+import requests
+
+# For a custom CA; if no custom CA is used, comment out this line.
+certifi.where = lambda: "/etc/pki/tls/cert.pem"
+
+# Configure logging
+hostname = socket.gethostname()
+logger = logging.getLogger("trinity-events")
+logging.basicConfig(
+    level=logging.INFO,
+    stream=sys.stderr,
+    format="%(asctime)s %(name)s %(levelname)s %(message)s ",
+)
+
+
+graphql_query = """
+    query QUERY_NAME ($first: Int, $last: Int, $after: String, $filter: EventFilter) {
+      events(first: $first, last: $last, after: $after, filter: $filter) {
+        pageInfo {
+          hasNextPage
+          endCursor
+        }
+        edges {
+          cursor
+          node {
+            id
+            actionTime
+            source
+            destination
+            sourcePort
+            destinationPort
+            transportProtocol
+            direction
+            trustInitiated
+            formulaMatches {
+                action {
+                    response
+                } 
+                formula {
+                  formulaId
+                  title
+                  background
+                  tags {
+                    category
+                    value
+                  }
+                }
+            }
+            applicationProtocol
+            applicationData {
+              __typename
+              ... on HttpRequestData {
+                method
+                path
+                host
+                userAgent
+              }
+              ... on HttpResponseData {
+                statusCode
+                statusString
+                server
+                contentType
+              }
+              ... on DnsData {
+                host
+              }
+              ... on TlsData {
+                sniHost
+              }
+            }
+          }
+        }
+      }
+    }
+    """
+
+
+class StreamOutput:
+    def __init__(self, output_config):
+        self.flatten = output_config["flatten"]
+        self.mapping = output_config.get("field_mapping", {})
+        format = output_config.get("format", "json")
+        if format == "json":
+            self.formatter = self.format_json
+        elif format == "leef":
+            self.flatten = True
+            self.formatter = self.format_leef
+
+    def output_event(self, event):
+        if self.flatten:
+            for flattened_event in flatten_event(event):
+                content = self.formatter(flattened_event)
+                print(content)
+        else:
+            content = self.formatter(event)
+            print(content)
+
+    def output_no_results(self):
```

Comment:
Ah, this was one of edits in the customer's copy.   It wasn't in the upstream script and I thought it only applied to S3.   

I'm not sure how this would work in LEEF.   I'm guessing the current customers that use the format won't want to see it.   Can make it a flag, I guess.

Reply:
I can see a `report-no-results` config option. What happens if we pass a dict with `time, status, result` to the LEEF formatter?

Reply:
546f4fd

Added to stdout JSON output.   Still need to add LEEF.